### PR TITLE
[SPARK-36686][SQL] Fix SimplifyConditionalsInPredicate to be null-safe

### PR DESCRIPTION
**Pull request by hypercubestart**
### What changes were proposed in this pull request?

fix SimplifyConditionalsInPredicate to be null-safe

Reproducible:

```
import org.apache.spark.sql.types.{StructField, BooleanType, StructType}
import org.apache.spark.sql.Row

val schema = List(
  StructField("b", BooleanType, true)
)
val data = Seq(
  Row(true),
  Row(false),
  Row(null)
)
val df = spark.createDataFrame(
  spark.sparkContext.parallelize(data),
  StructType(schema)
)

// cartesian product of true / false / null
val df2 = df.select(col("b") as "cond").crossJoin(df.select(col("b") as "falseVal"))
df2.createOrReplaceTempView("df2")

spark.sql("SELECT * FROM df2 WHERE IF(cond, FALSE, falseVal)").show()
// actual: 
// +-----+--------+                                                                
// | cond|falseVal|
// +-----+--------+
// |false|    true|
// +-----+--------+
spark.sql("SET spark.sql.optimizer.excludedRules=org.apache.spark.sql.catalyst.optimizer.SimplifyConditionalsInPredicate")
spark.sql("SELECT * FROM df2 WHERE IF(cond, FALSE, falseVal)").show()
// expected:
// +-----+--------+
// | cond|falseVal|
// +-----+--------+
// |false|    true|
// | null|    true|
// +-----+--------+
```

### Why are the changes needed?

is a regression that leads to incorrect results

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

existing tests

### CODE REVIEW

**Comment by hypercubestart** [comment]
cc for review: @cloud-fan @sigmod

**Comment by HyukjinKwon** [comment]
Can you add a test and show the reproducer with incorrect results in the PR description?

**Comment by HyukjinKwon** [comment]
Also, Apache Spark uses forked repository's resource. Please enable Github Actions in your forked repository (see also https://github.com/apache/spark/pull/33928/checks?check_run_id=3537007064)

**Comment by cloud-fan** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala [hunk 1]
```diff
@@ -56,16 +56,17 @@ object SimplifyConditionalsInPredicate extends Rule[LogicalPlan] {
     case And(left, right) => And(simplifyConditional(left), simplifyConditional(right))
     case Or(left, right) => Or(simplifyConditional(left), simplifyConditional(right))
     case If(cond, trueValue, FalseLiteral) => And(cond, trueValue)
-    case If(cond, trueValue, TrueLiteral) => Or(Not(cond), trueValue)
-    case If(cond, FalseLiteral, falseValue) => And(Not(cond), falseValue)
+    case If(cond, trueValue, TrueLiteral) => Or(Not(Coalesce(Seq(cond, FalseLiteral))), trueValue)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/SimplifyConditionalsInPredicate.scala:59]: good catch!

**Comment by cloud-fan** [comment]
The bug is pretty obvious, may not be worth an end-to-end test. It's still better to give a real query in the PR description to demonstrate the bug.

**Comment by hypercubestart** [comment]
> The bug is pretty obvious, may not be worth an end-to-end test. It's still better to give a real query in the PR description to demonstrate the bug.

thanks for review! added reproducible from real query that demonstrates the bug in the PR description

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.2!

**Pull request closed without merging**